python/cuml/neighbors/nearest_neighbors.pyx

```diff
@@ -393,5 +403,145 @@ class NearestNeighbors(Base):
+
+        else:
+            raise ValueError('Unsupported mode, must be one of "connectivity" '
+                             'or "distance" but got "%s" instead' % mode)
+
+        n_samples = indices.shape[0]
+        n_samples_fit = self.X_m.shape[0]
+        n_nonzero = n_samples * n_neighbors
+        rowptr = cp.arange(0, n_nonzero + 1, n_neighbors)
+        return cp.sparse.csr_matrix((distances, cp.ravel(indices), rowptr),
+                                    shape=(n_samples, n_samples_fit))
+
+
+def kneighbors_graph(X=None, n_neighbors=5, mode='connectivity', verbose=False,
+                     handle=None, algorithm="brute", metric="minkowski", p=2,
+                     include_self=False, metric_params=None):
+    """
+    Computes the (weighted) graph of k-Neighbors for points in X.
+
+    Parameters
+    ----------
+    X : array-like (device or host) shape = (n_samples, n_features)
+        Dense matrix (floats or doubles) of shape (n_samples, n_features).
+        Acceptable formats: cuDF DataFrame, NumPy ndarray, Numba device
+        ndarray, cuda array interface compliant array like CuPy
+
+    n_neighbors : Integer
+        Number of neighbors to search. If not provided, the n_neighbors
+        from the model instance is used (default=5)
+
+    mode : string (default='connectivity')
+        Values in connectivity matrix: 'connectivity' returns the
+        connectivity matrix with ones and zeros, 'distance' returns the
+        edges as the Euclidean distance between points
+
+    verbose : int or boolean (default = False)
+        Logging level
+
+    handle : cumlHandle
+        The cumlHandle resources to use
+
+    algorithm : string (default='brute')
+        The query algorithm to use. Currently, only 'brute' is supported.
+
+    metric : string (default='minkowski').
+        Distance metric to use. Supported distances are ['l1, 'cityblock',
+        'taxicab', 'manhattan', 'euclidean', 'l2', 'braycurtis', 'canberra',
+        'minkowski', 'chebyshev', 'jensenshannon', 'cosine', 'correlation']
+
+    p : float (default=2) Parameter for the Minkowski metric. When p = 1, this
+        is equivalent to manhattan distance (l1), and euclidean distance (l2)
+        for p = 2. For arbitrary p, minkowski distance (lp) is used.
+
+    include_self : bool or 'auto' (default=False)
+        Whether or not to mark each sample as the first nearest neighbor to
+        itself. If 'auto', then True is used for mode='connectivity' and False
+        for mode='distance'.
+
+    metric_params : dict, optional (default = None) This is currently ignored.
+
+    Returns
+    -------
+    A: sparse graph in CSR format, shape = (n_samples, n_samples_fit)
+        n_samples_fit is the number of samples in the fitted data where
+        A[i, j] is assigned the weight of the edge that connects i to k.
+        Values will be ones/zeros or Euclidean distance based on mode.
+
+    """
+    if not isinstance(X, NearestNeighbors):
```

Comment:
We should put this input option in the docs.

Reply:
Not sure if I understand, I meant to have this to catch users trying to use the "out the box" functionality by passing a NearestNeighbors instance as X which isn't how you should be using the function. I can remove it though if it's unnecessary?

Reply:
I was originally going to suggest that we scrap this feature altogether, however, I looked at Scikit-learn's API and found that it accepts a `BallTree` directly. As a result, I really don't think it's a bad idea to support it, but that support should either be documented or this line removed.

Reply:
That makes sense, I think since we only support `brute` right now I'll remove it, we'll have to make changes once we add `BallTree` anyways.